docs/site/Interceptors.md

```diff
@@ -0,0 +1,529 @@
+  }
+
+  // Bind options and provider for `NameValidator`
+  ctx.bind('valid-names').to(['John', 'Mary']);
+  ctx.bind('name-validator').toProvider(NameValidator);
+  ```
+
+### Method level interceptors
+
+A **public** static or prototype method on a class can be decorated with
+`@intercept` to attach interceptors to the target method. Please note
+interceptors don't apply to protected or private methods.
+
+#### Static methods
+
+```ts
+class MyControllerWithStaticMethods {
+  // Apply `log` to a static method
+  @intercept(log)
+  static async greetStatic(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  // Apply `log` to a static method with parameter injection
+  @intercept(log)
+  static async greetStaticWithDI(@inject('name') name: string) {
+    return `Hello, ${name}`;
+  }
+}
+```
+
+#### Prototype methods
+
+```ts
+class MyController {
+  // Apply `logSync` to a sync instance method
+  @intercept(logSync)
+  greetSync(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  // Apply `log` to a sync instance method
+  @intercept(log)
+  greet(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  // Apply `log` as a binding key to an async instance method
+  @intercept('log')
+  async greetWithABoundInterceptor(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  // Apply `log` to an async instance method with parameter injection
+  @intercept(log)
+  async greetWithDI(@inject('name') name: string) {
+    return `Hello, ${name}`;
+  }
+
+  // Apply `log` and `logSync` to an async instance method
+  @intercept('log', logSync)
+  async greetWithTwoInterceptors(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  // No interceptors are attached
+  async greetWithoutInterceptors(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  // Apply `convertName` to convert `name` arg to upper case
+  @intercept(convertName)
+  async greetWithUpperCaseName(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  // Apply `name-validator` backed by a provider class
+  @intercept('name-validator')
+  async greetWithNameValidation(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  // Apply `logError` to catch errors
+  @intercept(logError)
+  async greetWithError(name: string) {
+    throw new Error('error: ' + name);
+  }
+}
+```
+
+### Class level interceptors
+
+To apply interceptors to be invoked for all methods on a class, we can use
+`@intercept` to decorate the class. When a method is invoked, class level
+interceptors (if not explicitly listed at method level) are invoked before
+method level ones.
+
+```ts
+// Apply `log` to all methods on the class
+@intercept(log)
+class MyControllerWithClassLevelInterceptors {
+  // The class level `log` will be applied
+  static async greetStatic(name: string) {
+    return `Hello, ${name}`;
+  }
+
+  @intercept(log)
+  static async greetStaticWithDI(@inject('name') name: string) {
+    return `Hello, ${name}`;
+  }
+
+  // We can apply `@intercept` multiple times on the same method
```

Comment:
Please rework this example from a single code snippet with code comments into proper text with multiple code snippets.